cli/src/cmd/cast/storage.rs

```diff
@@ -0,0 +1,246 @@
+///
+/// https://github.com/ethereum/solidity/blob/develop/Changelog.md#065-2020-04-06
+const MIN_SOLC: Version = Version::new(0, 6, 5);
+
+#[derive(Debug, Clone, Parser)]
+pub struct StorageArgs {
+    // Storage
+    #[clap(help = "The contract address.", parse(try_from_str = parse_name_or_address), value_name = "ADDRESS")]
+    address: NameOrAddress,
+    #[clap(help = "The storage slot number (hex or decimal)", parse(try_from_str = parse_slot), value_name = "SLOT")]
+    slot: Option<H256>,
+    #[clap(long, env = "ETH_RPC_URL", value_name = "URL")]
+    rpc_url: Option<String>,
+    #[clap(
+        long,
+        short = 'B',
+        help = "The block height you want to query at.",
+        long_help = "The block height you want to query at. Can also be the tags earliest, latest, or pending.",
+        parse(try_from_str = parse_block_id),
+        value_name = "BLOCK"
+    )]
+    block: Option<BlockId>,
+
+    // Etherscan
+    #[clap(long, short, env = "ETHERSCAN_API_KEY", help = "etherscan API key", value_name = "KEY")]
+    etherscan_api_key: Option<String>,
+    #[clap(
+        long,
+        visible_alias = "chain-id",
+        env = "CHAIN",
+        help = "The chain ID the contract is deployed to.",
+        default_value = "mainnet",
+        value_name = "CHAIN"
+    )]
+    chain: Chain,
+
+    // Forge
+    #[clap(flatten)]
+    build: build::CoreBuildArgs,
+}
+
+impl StorageArgs {
+    pub async fn run(self) -> Result<()> {
+        let StorageArgs { address, block, build, rpc_url, slot, chain, etherscan_api_key } = self;
+
+        let rpc_url = consume_config_rpc_url(rpc_url);
+        let provider = try_get_http_provider(rpc_url)?;
+
+        let address = match address {
+            NameOrAddress::Name(name) => provider.resolve_name(&name).await?,
+            NameOrAddress::Address(address) => address,
+        };
+
+        // Slot was provided, perform a simple RPC call
+        if let Some(slot) = slot {
+            let cast = Cast::new(provider);
+            println!("{}", cast.storage(address, slot, block).await?);
+            return Ok(())
+        }
+
+        // No slot was provided
+        // Get deployed bytecode at given address
+        let address_code = provider.get_code(address, block).await?;
+        if address_code.is_empty() {
+            eyre::bail!("Provided address has no deployed code and thus no storage");
+        }
+
+        // Check if we're in a forge project
+        let project = build.project()?;
+        if project.paths.has_input_files() {
+            // Find in artifacts and pretty print
+            let project = with_storage_layout_output(project);
+            let out = compile(&project, false, false)?;
+            let match_code = |artifact: &ConfigurableContractArtifact| -> Option<bool> {
+                let bytes =
+                    artifact.deployed_bytecode.as_ref()?.bytecode.as_ref()?.object.as_bytes()?;
+                Some(bytes == &address_code)
```

Comment:
In the case of an external library, will the object be linked here? If it isn't then this match will always fail